python/frontend/cirrus/cirrus/GridSearch.py

```diff
@@ -173,63 +167,76 @@ def custodian(cirrus_objs, thread_id, num_jobs):
                     time.sleep(0.5)
                     start_time = time.time()
 
-        # Dictionary of commands per machine
-        command_dict = {}
-        for machine in self.machines:
-            command_dict[machine[0]] = []
 
-        # Grab commands each machine needs to run
-        for c in self.cirrus_objs:
-            c.get_command(command_dict)
-
-        # Write those commands into bash files
-        command_dict_to_file(command_dict)
-
-        # Number of threads
-        copy_threads = min(len(self.machines), self.num_jobs)
+        def unbuffer_instance(instance):
+            status, stdout, stderr = instance.buffer_commands(False)
+            if status != 0:
+                print("An error occurred while unbuffering commands on an"
+                      " instance. The exit code was %d and the stderr was:"
+                      % status)
+                print(stderr)
+                raise RuntimeError("An error occured while unbuffering"
+                                   " commands on an instance.")
 
-        # Copies bash files to machines and starts experiment
-        def copy_and_run(thread_id):
-            while True:
-                if thread_id >= len(self.machines):
-                    return
 
-                sh_file = "machine_%d.sh" % thread_id
-                ubuntu_machine = "ubuntu@%s" % self.machines[thread_id][0]
+        for instance in self.instances:
+            instance.buffer_commands(True)
 
-                cmd = "scp %s %s:~/" % (sh_file, ubuntu_machine)
-                os.system(cmd)
-                cmd = 'ssh %s "killall parameter_server; chmod +x %s; ./%s &"' % (ubuntu_machine, sh_file, sh_file)
-                os.system(cmd)
-                thread_id += copy_threads
-
-        p_lst = []
-        for i in range(copy_threads):
-            p = threading.Thread(target=copy_and_run, args=(i,))
-            p.start()
-            p_lst.append(p)
+        for cirrus_obj in self.cirrus_objs:
+            cirrus_obj.run(False)
 
-        [p.join() for p in p_lst]
+        threads = []
+        for instance in self.instances:
+            t = threading.Thread(
+                target=unbuffer_instance,
```

Comment:
Why do we create threads here that start right away to unbuffer the commands "queue" of experiment?

Reply:
Unbuffering an instance's command queue involves issuing all of the commands and waiting for their output. The threads let us unbuffer all of the instances' command queues in parallel, in order to reduce experiment startup time.